bool req_0
bool req_1
bool req_2
bool full_0
bool full_1